File: exercises/python_playwright/Shadow_DOM/shadow_dom_playwright.py
from playwright.sync_api import Page


class ShadowDOMPage:
    def __init__(self, page: Page):
        """Store page and locators used by the exercise."""
        self.page = page
        self.edit_field = page.locator("#editField")
        self.generate_button = page.locator("#buttonGenerate")
        self.copy_button = page.locator("#buttonCopy")

    def open(self) -> None:
        """Open the page over HTTPS despite default address being HTTP so clipboard functionality works."""
        self.page.goto("https://uitestingplayground.com/shadowdom")

    def generate_guid(self) -> str:
        """Generate a new GUID and return the input value."""
        self.generate_button.click()
        return self.edit_field.input_value()

    def copy_guid(self) -> None:
        """Copy the generated GUID to the clipboard."""
        self.copy_button.click()

File: exercises/javascript_playwright/shadow-dom/shadowDomPlaywright.js
export class ShadowDOMPage {
    /**
     *
     * @param {import('@playwright/test').Page} page
     */
    constructor(page) {
        this.page = page;
        this.guidField = page.locator("#editField");
        this.generateButton = page.locator("#buttonGenerate");
        this.copyButton = page.locator("#buttonCopy");
    }

    async openExercisePage() {
        await this.page.goto("https://uitestingplayground.com/shadowdom");
    }

    async clickGenerateGuidButton() {
        await this.generateButton.click();
    }

    async clickCopyButton() {
        await this.copyButton.click();
    }

    async readGuidField() {
        return await this.guidField.inputValue();
    }
}
